File: leetcode/js/add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let head = null;
    let tail = null;
    let carry = 0;

    while (l1 || l2) {
        let val1 = 0;
        let val2 = 0;
        let sum;
        let node;

        if (l1) {
            val1 = l1.val;
            l1 = l1.next;
        }

        if (l2) {
            val2 = l2.val;
            l2 = l2.next;
        }

        sum = val1 + val2 + carry;
        node = new ListNode(sum % 10);
        carry = sum >= 10 ? 1 : 0;

        if (head) {
            tail.next = node;
            tail = node;
        } else {
            head = tail = node;
        }
    }

    if (carry) {
        tail.next = new ListNode(1);
    }

    return head;
};

File: leetcode/java/AddTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 }

class AddTwoNumbers {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = null;
        ListNode tail = null;
        int carry = 0;

        while (l1 != null || l2 != null) {
            int val1 = 0;
            int val2 = 0;
            int sum;
            ListNode node;

            if (l1 != null) {
                val1 = l1.val;
                l1 = l1.next;
            }

            if (l2 != null) {
                val2 = l2.val;
                l2 = l2.next;
            }

            sum = val1 + val2 + carry;
            node = new ListNode(sum % 10);
            carry = sum >= 10 ? 1 : 0;

            if (head == null) {
                head = tail = node;
            } else {
                tail.next = node;
                tail = node;
            }
        }

        if (carry == 1) {
            tail.next = new ListNode(1);
        }

        return head;
    }
}
